packages/wallet/ui/src/util/keyManagement.js

```diff
@@ -0,0 +1,425 @@
+    },
+    BasicAllowance: {
+      typeUrl: '/cosmos.feegrant.v1beta1.BasicAllowance',
+    },
+  },
+});
+
+/**
+ * `/agoric.swingset.XXX` matches package agoric.swingset in swingset/msgs.go
+ */
+export const SwingsetMsgs = /** @type {const} */ ({
+  MsgWalletAction: {
+    typeUrl: '/agoric.swingset.MsgWalletAction',
+    aminoType: 'swingset/WalletAction',
+  },
+  MsgWalletSpendAction: {
+    typeUrl: '/agoric.swingset.MsgWalletSpendAction',
+    aminoType: 'swingset/WalletAction',
+  },
+});
+
+/**
+ * @typedef {{
+ *   owner: string, // base64 of raw bech32 data
+ *   action: string,
+ * }} WalletAction
+ * @typedef {{
+ *   owner: string, // base64 of raw bech32 data
+ *   spendAction: string,
+ * }} WalletSpendAction
+ */
+
+export const SwingsetRegistry = new Registry([
+  ...defaultRegistryTypes,
+  [SwingsetMsgs.MsgWalletAction.typeUrl, MsgWalletAction],
+  [SwingsetMsgs.MsgWalletSpendAction.typeUrl, MsgWalletSpendAction],
+]);
+
+/**
+ * TODO: estimate fee? use 'auto' fee?
+ * https://github.com/Agoric/agoric-sdk/issues/5888
+ *
+ * @returns {import('@cosmjs/stargate').StdFee}
+ */
+export const zeroFee = () => {
+  const { coinMinimalDenom: denom } = stableCurrency;
+  const fee = {
+    amount: [{ amount: '0', denom }],
+    gas: '100000', // TODO: estimate gas?
+  };
+  return fee;
+};
+
+/** @type {import('@cosmjs/stargate').AminoConverters} */
+const SwingsetConverters = {
+  // TODO: #3628, #4654
+  // '/agoric.swingset.MsgProvision': {
+  //   /* ... */
+  // },
+  [SwingsetMsgs.MsgWalletAction.typeUrl]: {
+    aminoType: SwingsetMsgs.MsgWalletAction.aminoType,
+    toAmino: ({ action, owner }) => ({
+      action,
+      owner: toBech32(bech32Config.bech32PrefixAccAddr, fromBase64(owner)),
+    }),
+    fromAmino: ({ action, owner }) => ({
+      action,
+      owner: toBase64(toAccAddress(owner)),
+    }),
+  },
+  [SwingsetMsgs.MsgWalletSpendAction.typeUrl]: {
+    aminoType: SwingsetMsgs.MsgWalletAction.aminoType,
+    toAmino: ({ spendAction, owner }) => ({
+      spendAction,
+      owner: toBech32(bech32Config.bech32PrefixAccAddr, fromBase64(owner)),
+    }),
+    fromAmino: ({ spendAction, owner }) => ({
+      spendAction,
+      owner: toBase64(toAccAddress(owner)),
+    }),
+  },
+};
+
+/**
+ * key for use in localStorage
+ *
+ * arbitrary; suffix chosen randomly to avoid collisions
+ */
+export const STORAGE_KEY = 'agoric.eis0Aigi';
+
+/**
+ * Maintain a key for signing non-spending messages in localStorage.
+ *
+ * See also `delegateWalletAction()` below.
+ *
+ * @param {object} io
+ * @param {typeof window.localStorage} io.localStorage
+ * @param {typeof import('@cosmjs/crypto').Random.getBytes} io.getBytes for key generation
+ */
+export const makeBackgroundSigner = async ({ localStorage, getBytes }) => {
+  const provideLocalKey = () => {
+    const stored = localStorage.getItem(STORAGE_KEY);
+    if (stored) {
+      return fromBase64(stored);
+    }
+    console.log(
```

Comment:
consider lowering the log level or removing.
```suggestion
    console.debug(
```